Refactor 'def set_config'

jzigmund opened the pull request:
The method ```set_config``` is used in a few Controllers#show and there is no needs to be a controller method, rather be helper's method as it prepares data for textual summaries and some specific actions under show action (i.e. devices, hardware info .. )

As a part of https://github.com/ManageIQ/manageiq/issues/13057

martinpovolny commented on app/helpers/host_helper.rb:
```diff
@@ -1,4 +1,11 @@
+    devices = get_devices
+    os_info = get_os_info
+    vmm_info = get_vmm_info
+    [devices, os_info, vmm_info]
```
I'd say that we don't need the temp variables here.

Also I'd say that the methods don't need the `get_` prefixes.

jzigmund replied:
yea I've created the PR just as a raw PR (or WIP) .. my point is to see review about the moving the code in to the helpers .. there are many things to improve in the code, but surely the code that makes your eyes bleeding, write it, I'll fix that ;-)

jzigmund commented:
according to Travis CI - some of the rspec tests are failing - marking PR as WIP

jzigmund commented:
Added 2 more commits, 1st is about a cleaning all ```set_config``` leftovers from HostController, 2nd is fixing the failing specs (previously 30 failing, now just 10), will continue tomorrow on the PR.

jzigmund commented:
updated PR, added commit to remove ```set_config``` from ApplicationController#set_summary_pdf_data, fixed rspecs, now the test should be green.
ToDo next:
- remove set_config from EmsClusterController and check other occurrences
- create tests for not covered code
- refactor new methods (after split the code to smaller methods), using of them in the views, fix rubocop issues

martinpovolny commented on app/helpers/textual_mixins/textual_devices.rb:
```diff
@@ -12,31 +12,37 @@ def textual_devices
+                                                                         :num_cores => @record.cpu_cores_per_socket}
+  end
+
+  def cpu_attributes
```
```
[
  [:cpu_type, _("CPU Type"), '', :processor]
  [:cpu_speed, _("CPU Speed"), _("Mhz"), :processor]
...
].each do 
```
Or you can wrap it into a class (Struct).

martinpovolny replied:
I think that this is better than adding some conditional logic below.

jzigmund replied:
yea that sounds good to me, i'm going to apply in the next iteration of the refactoring

jzigmund commented:
updated last commit, refactored ```get_devices``` (in progress working method name), please review and i will continue on other methods (os_info and vmm_info)

martinpovolny commented on spec/controllers/application_controller_spec.rb:
```diff
@@ -101,40 +101,6 @@
-
-    it "doesn't crash on nil filename" do
-      disk = FactoryGirl.create(:disk, :filename => nil, :controller_type => nil, :device_type => 'disk', :mode => "foo")
-      host_hardware = FactoryGirl.create(:hardware, :cpu_sockets => 2, :cpu_cores_per_socket => 4, :cpu_total_cores => 8, :disks => [disk])
```
This looks like we actually run some data through the `set_config`. I don't think we should just remove this. We should change this to call the new method and be more like a unit test.

jzigmund replied:
I agree, I've removed that as the method are not living in AppController anymore. I will add the tests for the newly created helpers' methods.

martinpovolny commented:
I'd prefer not to delete the tests, but addapt those. Other than that it looks good!

jzigmund commented:
@martinpovolny finished refactoring (+264 −368 so far so good :beers: ), please review.

Next: Add rspecs for the new methods

jzigmund commented:
added basic test examples for TextualDevice methods, created PR to main repo for new traits used in tests. Please review.

martinpovolny commented:
When ready for review, please, remove the WIP labels and texts. Thx!

jzigmund commented:
@martinpovolny please review, cannot remove WIP as it needs to have merged factories in main repo

martinpovolny commented:
> @martinpovolny please review, cannot remove WIP as it needs to have merged factories in main repo

As I said, once ready, please, remove the WIP flag. WIP PRs are typically not reviewed because reviewers filter PRs to review and exclude WIPs.

If you are blocked by backend/core PR then indicate that using the flag "pending core". That's a process we follow since the UI split.

martinpovolny commented:
Also, link the "pending core" PR in the description.

jzigmund commented:
@miq-bot remove_label pending core

martinpovolny commented on app/helpers/textual_mixins/textual_devices.rb:
```diff
@@ -0,0 +1,115 @@
+module TextualMixins::TextualDevices
+  Device = Struct.new(:name, :description, :units, :icon) do
```
Please, put the `Device` class into a separate file (as are other `Struct`-based classes).

Not only is it considered a best practice (separate file), but it also prevents all kinds of crazy class-loader issues when reloading files in dev mode.

jzigmund replied:
ok, sounds good .. going to move it to separate file

martinpovolny commented:
So the core repo PR that needed to be merged is already merged?

I see you remove the 'pending core' flag but have not commented on that.

Would be handy if you could link it in the description text as I requested. Then it would be easy for me or other reviewers to see its status.

martinpovolny commented:
Thx, @jzigmund. Hopefully one last thing: the Rubocops seem both right. Can you, please, fix those?

jzigmund commented:
@martinpovolny updated PR, moved ```Device``` Struct definition to the separated file, fixed Rubocop issue and fix building of network description, when the parts of final hash are nil(s)  https://github.com/ManageIQ/manageiq-ui-classic/pull/752/commits/c1d985293367c3c0958af7574a134638c0db0c10#diff-0477435a7227f32a30f48f81a7b52afcR89

Last Rubocop issue is wrong as we need to set nil there (in else branch) as the description always is initialized to the method name.

The main repo PR has been rejected as the devs recommened to use inline factories instead of the 
defining new ones. That PR has been already linked here (8 days ago) but seems that you missed that ?
![pr-link](https://cloud.githubusercontent.com/assets/2270962/24803138/90478a08-1baa-11e7-9b17-4a801cfe5241.png)

https://github.com/ManageIQ/manageiq/pull/14598 - again just for you ;-)

martinpovolny commented:
> That PR has been already linked here (8 days ago) but seems that you missed that ?

Yes, I missed that. I asked you to put it into the description (2 times was it?) so that I would not miss it. 

I am ofter wrong and I overlook things so I asked you to simplify my job of jumping between PRs and remembering the context by adding the information I need on the top of your PR.

Seems you decided not to do that and instead paste here a screnshot. I suppose it was to proof something? 

That is NOT COOL :-1: 

> Last Rubocop issue is wrong as we need to set nil there (in else branch) as the description always is initialized to the method name.

I don't see how the Rubocop issue is wrong. My understanding of the Ruby code matches what Rubocop wrote. Can you, please, demonstrate how we are both wrong (Robocop and me)? Or you can, please, fix what I asked you for.

Then I'd like to merge this PR finally.

Thx!

jzigmund commented:
@martinpovolny fixed the last Rubocop issue, the explanation of the issue was confusing to me, but I've checked that out in IRB and seen what's the point of that, tbh I've never met with that, I think it's not clear that it assigns nil for else part (which is not written in this case at all) - not sure why you get it that I'm saying you are wrong.

to the screenshot - intention was to navigate you to the references (as the GH doesn't give a possibility to direct link on that as in case of comments) - i've put the link directly under the screenshot again. I haven't added the link to the description as I wrote the core PR has been abandoned, and used inline factories in this PR, so it's no more related to this PR, but ofc I can add it if it's still desired

martinpovolny commented on app/helpers/textual_mixins/textual_devices.rb:
```diff
@@ -42,11 +42,11 @@ def disks_attributes
-                        :size            => size,
-                        :space           => prov}
+      device_name = _("Hard Disk (%{controller_type} %{location}), Size %{size},
+                      Percent Used Provisioned Space %{space}") % {:controller_type => ctrl_type,
```
@jzigmund, I am very unsure about what happens in the gettext catalog if you split a string across multiple lines. Are you sure whis will result in the same key as before? Or shall this be rather resolved with a backslash?

jzigmund replied:
yea sorry, it could lead to the trouble with fast_gettext, going to use backlash as you mentioned that.
